src/sql/src/plan/query.rs

```diff
@@ -3070,6 +3085,53 @@ fn plan_subscript_slice(
     .into())
 }
 
+fn plan_like(
+    ecx: &ExprContext,
+    expr: &Expr<Aug>,
+    pattern: &Expr<Aug>,
+    escape: Option<&Expr<Aug>>,
+    case_insensitive: bool,
+    not: bool,
+) -> Result<HirScalarExpr, PlanError> {
+    let ecx = ecx.with_name("LIKE argument");
+    let expr = plan_expr(&ecx, expr)?;
+    let haystack_type = ecx.scalar_type(&expr);
+    let haystack = match haystack_type {
+        Some(ScalarType::Char { length }) => HirScalarExpr::CallUnary {
+            func: UnaryFunc::PadChar(expr_func::PadChar { length }),
+            expr: Box::new(expr.type_as(&ecx, &haystack_type.unwrap())?),
+        },
+        _ => expr.type_as(&ecx, &ScalarType::String)?,
```

Comment:
Does this want to be `cast_to` instead of `type_as`?

Reply:
I don't have a good intuition for this yet. I'm worried in this case about a cast being too broad. How does one decide to use one over the other?

Reply:
Normally this gets handled by the function/operator selection code. You can essentially think of the function selection code as looking for any function overload where the concrete argument types can be implicitly cast to the parameter types. E.g., a function `like_with_escape(haystack: text, needle: text, escape: text)` would accept haystack, needle, and escape parameters of _any_ type that can be explicitly cast to `text`.

So we want to do the same thing here, and match what the function/operator selection code would do if this were a normal function call rather than a special bit of syntax. We might not have any tests for this, but I'll bet with the `type_as` instead of the `cast_to`, something like this will get incorrectly rejected:

```sql
select 'foo'::varchar like 'foo'::varchar;
```

Roughly speaking, implicit casts are a very safe and normal thing to do in SQL. If an implicit cast exists between two types, we're meant to treat them as roughly equivalent; they're the kind of cast that usually succeeds. E.g., taking a `varchar` to `text`, or taking `date` to a `timestamp`.

Reply:
Thanks for the explanation! Fixed.